File: packages/python/src/vendorval_sdk/resources/_certifications.py
"""Certifications resource (sync + async). Phase N customer-facing
reshape, Workstream B.

Today this surface is read-only — `list` + `retrieve`. POST + DELETE
(manual upload + revoke) land in a follow-up SDK release once those
API routes ship.
"""

from __future__ import annotations

from typing import Any
from urllib.parse import quote

import httpx

from .._models import Response
from .._request import ResolvedConfig, execute_async, execute_sync, prepare


def _build_query(
    *,
    entity_id: str | None,
    issuer: str | None,
    status: str | None,
    expiring_within_days: int | None,
    limit: int | None,
    offset: int | None,
) -> dict[str, Any]:
    return {
        "entity_id": entity_id,
        "issuer": issuer,
        "status": status,
        "expiring_within_days": expiring_within_days,
        "limit": limit,
        "offset": offset,
    }


class CertificationsResource:
    def __init__(self, cfg: ResolvedConfig, client: httpx.Client) -> None:
        self._cfg = cfg
        self._client = client

    def list(
        self,
        *,
        entity_id: str | None = None,
        issuer: str | None = None,
        status: str | None = None,
        expiring_within_days: int | None = None,
        limit: int | None = None,
        offset: int | None = None,
    ) -> Response:
        """List certifications for the calling org.

        Returns the full list envelope verbatim so callers see pagination
        metadata (`total`, `has_more`, `limit`, `offset`) without
        re-querying for the count. Access rows via `response["data"]`,
        or call `response.to_dict()` to work with the full payload as a
        plain dictionary.
        """
        prepared = prepare(
            self._cfg,
            method="GET",
            path="/v1/certifications",
            query=_build_query(
                entity_id=entity_id,
                issuer=issuer,
                status=status,
                expiring_within_days=expiring_within_days,
                limit=limit,
                offset=offset,
            ),
        )
        res = execute_sync(self._client, prepared)
        return Response(res.data, res.request_id, res.status)

    def retrieve(self, certification_id: str) -> Response:
        """Fetch a single certification by its public id (`cert_…`)."""
        prepared = prepare(
            self._cfg,
            method="GET",
            path=f"/v1/certifications/{quote(certification_id, safe='')}",
        )
        res = execute_sync(self._client, prepared)
        return Response(res.data, res.request_id, res.status)


class AsyncCertificationsResource:
    def __init__(self, cfg: ResolvedConfig, client: httpx.AsyncClient) -> None:
        self._cfg = cfg
        self._client = client

    async def list(
        self,
        *,
        entity_id: str | None = None,
        issuer: str | None = None,
        status: str | None = None,
        expiring_within_days: int | None = None,
        limit: int | None = None,
        offset: int | None = None,
    ) -> Response:
        prepared = prepare(
            self._cfg,
            method="GET",
            path="/v1/certifications",
            query=_build_query(
                entity_id=entity_id,
                issuer=issuer,
                status=status,
                expiring_within_days=expiring_within_days,
                limit=limit,
                offset=offset,
            ),
        )
        res = await execute_async(self._client, prepared)
        return Response(res.data, res.request_id, res.status)

    async def retrieve(self, certification_id: str) -> Response:
        prepared = prepare(
            self._cfg,
            method="GET",
            path=f"/v1/certifications/{quote(certification_id, safe='')}",
        )
        res = await execute_async(self._client, prepared)
        return Response(res.data, res.request_id, res.status)

File: packages/node/src/resources/certifications.ts
import { performRequest, type ResolvedClientOptions } from "../request.js";
import type {
  Certification,
  CertificationsListParams,
  CertificationsListResponse,
} from "../types/shared.js";

/**
 * `client.certifications.*` — read access to entity credentials
 * (state MWBE, NMSDC, WBENC, ISO, SOC 2, etc.). Phase N customer-facing
 * reshape, Workstream B.
 *
 * Today this surface is read-only; POST + DELETE (manual upload + revoke)
 * land in a follow-up SDK release once those API routes ship.
 */
export class CertificationsResource {
  constructor(private readonly client: ResolvedClientOptions) {}

  /**
   * List certifications for the calling org. Filters narrow the result
   * set on the server. Returns the full list envelope verbatim so
   * callers see pagination metadata (`total`, `has_more`, `limit`,
   * `offset`) without re-querying for the count.
   *
   * Async iteration helper:
   *
   *   for (const cert of (await client.certifications.list()).data) { … }
   *
   * Cursor pagination is a planned follow-up — when it lands, this
   * shape will accept a `cursor` param + return `next_cursor` without
   * a breaking change.
   */
  async list(
    params: CertificationsListParams = {},
  ): Promise<CertificationsListResponse & { _requestId: string | null }> {
    const query: Record<string, string | number | boolean | undefined> = {};
    if (params.entity_id !== undefined) query.entity_id = params.entity_id;
    if (params.issuer !== undefined) query.issuer = params.issuer;
    if (params.status !== undefined) query.status = params.status;
    if (params.expiring_within_days !== undefined) {
      query.expiring_within_days = params.expiring_within_days;
    }
    if (params.limit !== undefined) query.limit = params.limit;
    if (params.offset !== undefined) query.offset = params.offset;

    const res = await performRequest<CertificationsListResponse>(this.client, {
      method: "GET",
      path: "/v1/certifications",
      query,
    });

    return { ...res.data, _requestId: res.requestId };
  }

  /**
   * Fetch a single certification by its public id (`cert_…`).
   * Throws a NotFound error if the id doesn't resolve under the
   * caller's org.
   */
  async retrieve(id: string): Promise<Certification & { _requestId: string | null }> {
    const res = await performRequest<Certification>(this.client, {
      method: "GET",
      path: `/v1/certifications/${encodeURIComponent(id)}`,
    });
    return { ...res.data, _requestId: res.requestId };
  }
}
